src/main/java/com/google/devtools/build/lib/analysis/producers/PlatformFlagsProducer.java

```diff
@@ -100,7 +100,12 @@ private StateMachine parsePlatformFlags(Tasks tasks) {
 
     PackageContext packageContext =
         PackageContext.of(platformLabel.getPackageIdentifier(), this.platformRepositoryMapping);
-    ParsedFlagsValue.Key parsedFlagsKey = ParsedFlagsValue.Key.create(flags, packageContext);
+    ParsedFlagsValue.Key parsedFlagsKey =
+        ParsedFlagsValue.Key.create(
+            flags,
+            packageContext,
+            // Include default values so that any flags explicitly reset to the default are kept.
+            /* includeDefaultValues= */ true);
```

Comment:
This is a good point: the code at present leaves the defaults for Starlark flags in the build options.

I think what is needed is for `ParsedFlagsFunction` to indicate that some flags get set to a specific value, and that some are removed entirely. I'm not sure how to model that: possibly with an `Optional`?

> I think I understand the bug though. The code as is doesn't even register flags being set to their defaults as something that needs to be changed, since the custom StarlarkOptionsParser just drops them?

Correct, before this change, if the platform says `--//starlark:flag=default`, it just isn't present in the list of flags. So something needs to signal "this flag needs to be explicitly removed".

Reply:
I'm curious what @aranguyen makes of this since I think scoped transitions will have similar issues.

Which code uses the results of `PlatformFlagsProducer` to produce the final `BuildOptions` that feed the configuration?  Is that somewhere in `BuildConfigurationKeyProducer`?